hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisOverReplicationHandler.java

```diff
@@ -0,0 +1,247 @@
+        sortReplicas(replicaCount.getReplicas());
+
+    // retain one replica per unique origin datanode if the container is not
+    // closed
+    final Map<UUID, ContainerReplica> uniqueReplicas =
+        new LinkedHashMap<>();
+    if (replicaCount.getContainer().getState() !=
+        HddsProtos.LifeCycleState.CLOSED) {
+      eligibleReplicas.stream()
+          // get replicas with state that matches container state
+          .filter(r -> ReplicationManager.compareState(
+              replicaCount.getContainer().getState(),
+              r.getState()))
+          .forEach(r -> uniqueReplicas
+              .putIfAbsent(r.getOriginDatanodeId(), r));
+
+      // note that this preserves order of the List
+      eligibleReplicas.removeAll(uniqueReplicas.values());
+    }
+
+    Set<DatanodeDetails> pendingDeletion = new HashSet<>();
+    // collect the DNs that are going to have their container replica deleted
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        pendingDeletion.add(op.getTarget());
+      }
+    }
+
+    // replicas that are not on IN_SERVICE nodes or are already pending
+    // delete are not eligible
+    // TODO what about nodes that are not healthy?
+    eligibleReplicas.removeIf(
+        replica -> replica.getDatanodeDetails().getPersistedOpState() !=
+            HddsProtos.NodeOperationalState.IN_SERVICE ||
+            pendingDeletion.contains(replica.getDatanodeDetails()));
+
+    return eligibleReplicas;
+  }
+
+  /**
+   * Sorts replicas using {@link ContainerReplica#hashCode()} (ContainerID and
+   * DatanodeDetails).
+   * @param replicas replicas to sort
+   * @return sorted List
+   */
+  private List<ContainerReplica> sortReplicas(
+      Collection<ContainerReplica> replicas) {
+    return replicas.stream()
+        .sorted(Comparator.comparingLong(ContainerReplica::hashCode))
+        .collect(Collectors.toList());
+  }
+
+  private Map<DatanodeDetails, SCMCommand<?>> createCommands(
+      ContainerInfo containerInfo, List<ContainerReplica> replicas,
+      int excess) {
+    Map<DatanodeDetails, SCMCommand<?>> commands = new HashMap<>();
+
+    /*
+    Over replication means we have enough healthy replicas, so unhealthy
+    replicas can be deleted. This might make the container violate placement
+    policy.
+     */
+    List<ContainerReplica> unhealthyReplicas = new ArrayList<>();
+    for (ContainerReplica replica : replicas) {
+      if (excess == 0) {
+        return commands;
+      }
+      if (!ReplicationManager.compareState(
+          containerInfo.getState(), replica.getState())) {
+        commands.put(replica.getDatanodeDetails(),
+            createDeleteCommand(containerInfo));
+        unhealthyReplicas.add(replica);
+        excess--;
+      }
+    }
+    replicas.removeAll(unhealthyReplicas);
+
+    /*
+    Remove excess replicas if that does not make the container mis replicated.
```

Comment:
We are hopefully going to improve this area, by asking the placement policy to tell us which replicas to remove and keep the placement correct. However that is not done yet, so for now best to keep this as it is.